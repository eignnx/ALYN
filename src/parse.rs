use lalrpop_util::lalrpop_mod;

lalrpop_mod!(
    #[allow(clippy::ptr_arg)]
    #[rustfmt::skip]
    grammar
);

pub use grammar::*;

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    macro_rules! test_parse {
        ($rule:ident, $src:expr) => {
            assert_debug_snapshot!(grammar::$rule::new().parse("<test>", $src));
        };
        ($rule:ident, $filename:expr, $src:expr) => {
            assert_debug_snapshot!(grammar::$rule::new().parse($filename, $src));
        };
    }

    #[test]
    fn bit_and() {
        test_parse!(RValParser, "x & y");
    }

    #[test]
    fn addr_deref_call() {
        test_parse!(RValParser, "&*x()");
    }

    #[test]
    fn pluses_and_minuses() {
        test_parse!(
            RValParser,
            "
            x - -y + z
        "
        );
    }

    #[test]
    fn operator_precedence() {
        test_parse!(
            RValParser,
            "
            x & -y + -z == 1 - 2 | 3
        "
        );
    }

    #[test]
    fn big_long_term() {
        test_parse!(
            RValParser,
            "
            *x + (127b - (+324)) - (-4) == **qwerty123 + asdf() - zxcv(1, 2)
        "
        );
    }

    #[test]
    fn type_transmutation_cast() {
        test_parse!(RValParser, "1 + 2 & bitcast{nat}(-3) & 4 + 5");
    }

    #[test]
    fn bitcast_with_nested_expr() {
        test_parse!(RValParser, "bitcast{nat}(12b + 40b)");
    }

    #[test]
    fn assignment() {
        test_parse!(StmtParser, "*lhs = 123 + *ptr;");
    }

    #[test]
    fn let_stmt() {
        test_parse!(StmtParser, "let lhs = 123 + *ptr;");
    }

    #[test]
    fn if_stmt() {
        test_parse!(StmtParser, "if 1 == 2 { lhs = 123; } else { ret asfd(); }");
    }

    #[test]
    fn simple_subr() {
        test_parse!(SubrDeclParser, "subr blah() { ret; }");
    }

    #[test]
    fn sumfac() {
        test_parse!(
            SubrDeclParser,
            "
            subr sumfac(n: nat) nat {
                ret n + sumfac(n - 1);
            }
        "
        );
    }

    #[test]
    fn main_with_args() {
        test_parse!(SubrDeclParser, "subr main(args: **byte) { ret; }");
    }

    #[test]
    fn min_subr() {
        test_parse!(
            SubrDeclParser,
            "
            subr min(a: int, b: int) int {
                if a > b {
                    ret b;
                } else {
                    ret a;
                }
            }
        "
        );
    }

    #[test]
    fn mathy_module() {
        test_parse!(
            ModuleParser,
            "mathy.alyn",
            "
            subr min(a: int, b: int) int {
                if a > b {
                    ret b;
                } else {
                    ret a;
                }
            }

            subr fib(n: nat) nat {
                if n < 2 {
                    ret 1;
                } else {
                    ret fib(n - 1) + fib(n - 2);
                }
            }
        "
        );
    }
}
