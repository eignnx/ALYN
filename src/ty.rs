use std::fmt;

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Ty {
    Void,
    Bool,
    Byte,
    Nat,
    Int,
    Ptr(Box<Ty>),
    Subr(Vec<Ty>, Box<Ty>),
}

impl fmt::Display for Ty {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Ty::Void => write!(f, "void"),
            Ty::Bool => write!(f, "bool"),
            Ty::Byte => write!(f, "byte"),
            Ty::Nat => write!(f, "nat"),
            Ty::Int => write!(f, "int"),
            Ty::Ptr(ty) => write!(f, "*{ty}"),
            Ty::Subr(param_tys, ret_ty) => {
                write!(f, "subr(")?;
                for (i, param_ty) in param_tys.iter().enumerate() {
                    if i > 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{param_ty}")?;
                }
                write!(f, ")")?;
                if !matches!(ret_ty.as_ref(), Ty::Void) {
                    write!(f, " -> {ret_ty}")?;
                }
                Ok(())
            }
        }
    }
}
