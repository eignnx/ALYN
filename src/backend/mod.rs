use crate::{canon, ir};

pub mod lark;

pub trait Backend {
    type Temporary;
    type Instruction;
    fn stmt_to_asm(&mut self, stmt: canon::Stmt);
    fn expr_to_asm(&mut self, rval: canon::RVal, opt_dst: impl Into<Option<Self::Temporary>>) -> Self::Temporary;
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    fn compile_to_lark(src: &str) -> Vec<lark::Instr> {
        use crate::parse;
        let mut mod_ast = parse::ModuleParser::new().parse("<test>", src).unwrap();
        use crate::ty::Ty;
        let mut tcx = crate::tcx::Tcx::new(Ty::Void);
        mod_ast.check_ty(&mut tcx);

        let mut out = vec![];
        let mut be = lark::LarkBackend::new(&mut out);
        for decl in mod_ast.decls {
            let subr_lbl = crate::names::Lbl::SubrStart(decl.value.name);
            let ir_stmts = decl.value.to_ir().into_iter().map(|w| w.as_stmt()).collect();
            let stmts = canon::canonicalize(subr_lbl, ir_stmts);
            for stmt in stmts {
                be.stmt_to_asm(stmt);
            }
        }
        be.render().to_vec()
    }

    #[test]
    fn basic() {
        assert_debug_snapshot!(compile_to_lark(
        "
            subr main() nat {
                ret 0;
            }
        "
        ));
    }

    #[test]
    fn binop() {
        assert_debug_snapshot!(compile_to_lark(
            "
            subr main() nat {
                let a = 24;
                ret 1 + a - 49;
            }
        "
        ));
    }

    #[test]
    fn simple_if() {
        assert_debug_snapshot!(compile_to_lark(
            "
            subr main() nat {
                if 1 == 1 {
                    let x = 1;
                } else {
                    let x = 2;
                }
            }
        "
        ));
    }

    #[test]
    fn simple_while() {
        assert_debug_snapshot!(compile_to_lark(
            "
            subr main() nat {
                let x = 1;
                while x < 10 {
                    x = x + 1;
                }
            }
        "
        ));
    }

    #[test]
    fn nested_call() {
        assert_debug_snapshot!(compile_to_lark(
            "
            subr add3(a: nat, b: nat, c: nat) nat {
                ret a + b + c;
            }

            subr main() nat {
                ret add3(1, add3(9, 8, 7), 2 + 3);
            }
        "
        ));
    }
}
