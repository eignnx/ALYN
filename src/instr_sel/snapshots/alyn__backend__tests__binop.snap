---
source: src/backend/mod.rs
expression: "compile_to_lark(\"\n            subr main() nat {\n                let a = 24;\n                ret 1 + a - 49;\n            }\n        \")"
---
[
    subr<main>:,
    li %a, 24,
    li %li_nat.0, 1,
    subi %imm_binop_res.1, %a, 49,
    add $rv, %li_nat.0, %imm_binop_res.1,
    jr $ra,
]
